### PR TITLE
+ new_grouped_df() function.

### PR DESCRIPTION
**Pull request by romainfrancois**
This might need some work on the documentation. Added the `new_grouped_df`: 

``` r
library(dplyr)

tbl <- new_grouped_df(
  tibble(x = rnorm(10)),
  tibble(.rows = replicate(5, sample(1:10, replace = TRUE), simplify = FALSE))
)
summarise(tbl, x = mean(x))
#> # A tibble: 1 x 1
#>        x
#>    <dbl>
#> 1 -0.271
```

`new_grouped_df` tests that all the indices are within the bounds. The `class` argument can be used to add an additional class, and that should be used to disable methods that don't make sense. 

I guess it's not supposed to be used directly by the end user,  but rather by 📦 developers who do alternative groupings.

### CODE REVIEW

File: R/grouped-df.r [hunk 1]
```diff
@@ -26,6 +26,30 @@ grouped_df <- function(data, vars, drop) {
   grouped_df_impl(data, unname(vars))
 }
 
+#' Create a custom grouped tibble
+#'
+#' @param .data A data frame
+#' @param group_data The grouped structure, see details
+#' @param class An additional class for the resulting tibble
+#'
```
Comment (hadley) [inline R/grouped-df.r]: Should be `@keywords internal`

File: R/grouped-df.r [hunk 2]
```diff
@@ -26,6 +26,30 @@ grouped_df <- function(data, vars, drop) {
...
+#' # 5 bootstrap samples
+#' tbl <- new_grouped_df(
+#'   tibble(x = rnorm(10)),
+#'   tibble(.rows = replicate(5, sample(1:10, replace = TRUE), simplify = FALSE))
+#' )
+#' # mean of each bootstrap sample
+#' summarise(tbl, x = mean(x))
+#'
+#' @export
+new_grouped_df <- function(.data, group_data, class = NULL) {
```
Comment (hadley) [inline R/grouped-df.r]: If we're going to make this subclassable, we need `...` to be passed on. I don't think subclassing can fully work until we depend on vctrs, but I think it's ok to provide partial support and complete it in the next release.

File: R/grouped-df.r [hunk 3]
```diff
@@ -26,6 +26,30 @@ grouped_df <- function(data, vars, drop) {
   grouped_df_impl(data, unname(vars))
 }
 
+#' Create a custom grouped tibble
+#'
+#' @param .data A data frame
+#' @param group_data The grouped structure, see details
```
Comment (hadley) [inline R/grouped-df.r]: I think you should inline the details here

File: R/grouped-df.r [hunk 4]
```diff
@@ -26,6 +26,30 @@ grouped_df <- function(data, vars, drop) {
   grouped_df_impl(data, unname(vars))
 }
 
+#' Create a custom grouped tibble
+#'
```
Comment (hadley) [inline R/grouped-df.r]: A low-level constructor for the `grouped_tbl` class. This constructor is designed to be high-performance so only check types, not values. This means it is the callers responsibility to create valid values, and hence this is for expert use only.

File: R/grouped-df.r [hunk 5]
```diff
@@ -26,6 +26,30 @@ grouped_df <- function(data, vars, drop) {
...
+#' tbl <- new_grouped_df(
+#'   tibble(x = rnorm(10)),
+#'   tibble(.rows = replicate(5, sample(1:10, replace = TRUE), simplify = FALSE))
+#' )
+#' # mean of each bootstrap sample
+#' summarise(tbl, x = mean(x))
+#'
+#' @export
+new_grouped_df <- function(.data, group_data, class = NULL) {
+  new_grouped_df_impl(.data, group_data, class)
```
Comment (hadley) [inline R/grouped-df.r]: `stopifnot(is.data.frame(.data))`?

File: R/grouped-df.r (hunk 1 above)
Comment (romainfrancois) [reply R/grouped-df.r]: ✅

File: R/grouped-df.r (hunk 3 above)
Comment (romainfrancois) [reply R/grouped-df.r]: ✅

File: R/grouped-df.r (hunk 4 above)
Comment (romainfrancois) [reply R/grouped-df.r]: In that case I should remove the bounds checks.

File: R/grouped-df.r [hunk 6]
```diff
@@ -26,6 +26,39 @@ grouped_df <- function(data, vars, drop) {
...
+#'
+#' @keywords internal
+#' @export
+new_grouped_df <- function(.data, groups, class = c("grouped_df", "tbl_df", "tbl", "data.frame"), ...) {
+  stopifnot(
+    is.data.frame(.data),
+    is.data.frame(groups),
+    tail(names(groups), 1L) == ".rows"
+  )
+  structure(.data, groups = groups, class = class, ... )
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply R/grouped-df.r]: Thanks. 🤹‍♂️ too many branches at the moment, but yeah that feels right.

File: R/grouped-df.r (hunk 4 above)
Comment (hadley) [reply R/grouped-df.r]: Yeah maybe also export `validate_grouped_df()`?

File: R/grouped-df.r [hunk 7]
```diff
@@ -26,6 +26,44 @@ grouped_df <- function(data, vars, drop) {
...
+  stopifnot(
+    is.data.frame(.data),
+    is.data.frame(groups),
+    tail(names(groups), 1L) == ".rows"
+  )
+  structure(
+    .data,
+    groups = groups,
+    ...,
+    class = c(class, c("grouped_df", "tbl_df", "tbl", "data.frame"))
```
Comment (hadley) [inline R/grouped-df.r]: Extra `c`

File: R/grouped-df.r (hunk 4 above)
Comment (romainfrancois) [reply R/grouped-df.r]: ✅

**Pull request merged**